noir-projects/noir-protocol-circuits/crates/types/src/merkle_tree/variable_merkle_tree.nr

```diff
@@ -16,6 +16,35 @@ unconstrained fn get_height(input: u32, start: u8) -> u8 {
     height
 }
 
+/// Calculates the previous power of 2.
+///
+/// The previous power of 2 is the largest power of 2 that is less than or equal to a number.
+/// For example:
+/// - For 7, previous power of 2 is 4 (2^2)
+/// - For 10, previous power of 2 is 8 (2^3)
+/// - For 16, previous power of 2 is 16 (2^4)
+///
+/// # Arguments
+/// * value - Integer value for which we need to find the previous power of 2
+///
+/// # Returns
+/// The previous power of 2
+fn get_prev_power_2(value: u32) -> u32 {
+    // Find size of tree required
+    let height = unsafe {
+        //@safety This is safe because we derive next and previous powers of 2 below based on the height
+        // and the powers are then checked against the input value.
+        get_height(value, 0)
+    };
+
+    let next_power_2 = 2 << height;
+    let prev_power_2 = next_power_2 / 2;
+    assert((value == 0) | (value == 1) | (value > prev_power_2));
+    assert(value <= next_power_2);
```

Comment:
```suggestion
    let next_power_exponent = unsafe {
        //@safety This is a hint that we'll use to compute the next and previous powers of two, which we check to be
        // larger and smaller than respectively. The get_height function happens to return exactly what we need.
        get_height(value, 0)
    };

    let next_power_2 = 2 << next_power_exponent;
    let prev_power_2 = next_power_2 / 2;
    assert((value == 0) | (value == 1) | (value > prev_power_2));
    assert(value <= next_power_2);
```

This is more readable I think. Also this is odd - it seems like `prev_power_2` is almost complete unconstrained if `value` is 0 or 1, it can be any value except 0.

Reply:
Since the underconstrained bug became real I decided to move the changes relate to `VariableMerkleTree` to [this stack of PRs ](https://github.com/AztecProtocol/aztec-packages/pull/11165). I included Tom's optimization in the upper one.